# TFM - Grupo 2 - Entrega 18/05/2024

In [ ]:
!pip install ydata_profiling
#%pip install -q ipywidgets
#!conda install -c conda-forge ipywidgets

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
import datetime
import ipywidgets

#from ydata_profiling import ProfileReport

%matplotlib inline
pd.set_option('display.max_columns', None)

# TASK 1 - EAD (EXPLORATORY DATA ANALYSIS)

dataset daily_calendar_with_events.csv

In [4]:
df1 = pd.read_csv("data/daily_calendar_with_events.csv")

In [5]:
df1.head()

date    weekday  weekday_int    d event
0  2011-01-29   Saturday            1  d_1   NaN
1  2011-01-30     Sunday            2  d_2   NaN
2  2011-01-31     Monday            3  d_3   NaN
3  2011-02-01    Tuesday            4  d_4   NaN
4  2011-02-02  Wednesday            5  d_5   NaN

In [6]:
df1['date'] = pd.to_datetime(df1.date)
df1.set_index('date', inplace=True)

In [7]:
df1.head(1)

weekday  weekday_int    d event
date                                        
2011-01-29  Saturday            1  d_1   NaN

In [8]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1913 entries, 2011-01-29 to 2016-04-24
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   weekday      1913 non-null   object
 1   weekday_int  1913 non-null   int64 
 2   d            1913 non-null   object
 3   event        26 non-null     object
dtypes: int64(1), object(3)
memory usage: 74.7+ KB


In [9]:
# Substituir linhas com valores NaN por 'Normalday'

df1.fillna('Normalday', inplace=True)

In [10]:
df1.head()

weekday  weekday_int    d      event
date                                              
2011-01-29   Saturday            1  d_1  Normalday
2011-01-30     Sunday            2  d_2  Normalday
2011-01-31     Monday            3  d_3  Normalday
2011-02-01    Tuesday            4  d_4  Normalday
2011-02-02  Wednesday            5  d_5  Normalday

In [11]:
df1.shape

(1913, 4)

In [12]:
df1['event'].value_counts()

Normalday         1887
SuperBowl            6
Ramadan starts       5
Thanksgiving         5
NewYear              5
Easter               5
Name: event, dtype: int64

In [13]:
# Dados em falta

df1.isnull().sum()

weekday        0
weekday_int    0
d              0
event          0
dtype: int64

Dataset item_prices.csv

In [14]:
#profile = ProfileReport(df1, title="Pandas Profiling Report")

In [15]:
#profile.to_file("daily_calendar_with_events.html")
#profile

In [16]:
df2 = pd.read_csv("data/item_prices.csv")

df2.info()
df2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6965706 entries, 0 to 6965705
Data columns (total 5 columns):
 #   Column      Dtype  
---  ------      -----  
 0   item        object 
 1   category    object 
 2   store_code  object 
 3   yearweek    float64
 4   sell_price  float64
dtypes: float64(2), object(3)
memory usage: 265.7+ MB


item    category store_code  yearweek  sell_price
0  ACCESORIES_1_001  ACCESORIES      NYC_1  201328.0     12.7414
1  ACCESORIES_1_001  ACCESORIES      NYC_1  201329.0     12.7414
2  ACCESORIES_1_001  ACCESORIES      NYC_1  201330.0     10.9858
3  ACCESORIES_1_001  ACCESORIES      NYC_1  201331.0     10.9858
4  ACCESORIES_1_001  ACCESORIES      NYC_1  201332.0     10.9858

Em um contexto de uma tabela de preços de itens (item_prices), o valor de yearweek geralmente seria usado para indicar em qual semana específica os preços foram registrados ou atualizados para um determinado item. Aqui está uma explicação mais detalhada:

Identificação da Semana: O valor yearweek provavelmente consistiria em um número que combina o ano com o número da semana dentro desse ano. Por exemplo, "201328" pode significar a vigésima oitava semana do ano de 2013.

In [17]:
# df2.drop('yearweek', axis=1, inplace=True) # Drop da coluna 'yearweek' 

In [18]:
# Dados em falta

df2.isnull().sum()

item               0
category           0
store_code         0
yearweek      243920
sell_price         0
dtype: int64

In [19]:
# Eliminar linhas duplicadas

df2 = df2.drop_duplicates()

In [20]:
# Eliminar linhas com NANs

df2 = df2.dropna()

In [21]:
df2.isnull().sum()

item          0
category      0
store_code    0
yearweek      0
sell_price    0
dtype: int64

In [22]:
df2. info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6721786 entries, 0 to 6965697
Data columns (total 5 columns):
 #   Column      Dtype  
---  ------      -----  
 0   item        object 
 1   category    object 
 2   store_code  object 
 3   yearweek    float64
 4   sell_price  float64
dtypes: float64(2), object(3)
memory usage: 307.7+ MB


In [23]:
#profile = ProfileReport(df2, title="Pandas Profiling Report")

In [24]:
#profile.to_file("item_prices.html")
#profile

DataSet item_sales.csv

In [ ]:
df3 = pd.read_csv("data/item_sales.csv")

df3.head()

In [ ]:
df3.shape

In [ ]:
df3.info()

In [ ]:
plt.figure(figsize=(10, 4), dpi=80)
sns.histplot(data=df3, x="region", shrink=.8, stat='percent')
plt.title("Gráfico de barras - region")
plt.show()

In [ ]:
plt.figure(figsize=(15, 4), dpi=80)
sns.histplot(data=df3, x="department", shrink=.8, stat='percent')
plt.title("Gráfico de barras - department")
plt.show()

In [ ]:
df3.describe()

In [ ]:
df3.describe(include=[object])

In [ ]:
df3['region'].value_counts()

In [ ]:
df3['department'].value_counts()

In [ ]:
df3['store'].value_counts()

In [ ]:
df3['category'].value_counts()

In [ ]:
# para excluir todas as colunas numéricas de uma vez
#df3_sem_numericas = df3.select_dtypes(exclude=['number'])
#df3_sem_numericas


In [ ]:
#df3_sem_numericas.shape

In [ ]:
#profile = ProfileReport(df3_sem_numericas, title="Pandas Profiling Report")

In [ ]:
#profile.to_file("item_sales.html")
#profile

In [ ]:
sales_melted = df3.melt(id_vars=['id', 'item', 'category', 'department', 'store', 'store_code', 'region'], var_name='d', value_name='value_sales')

# Juntar os datasets com base na coluna 'd'
merged_df = pd.merge(sales_melted, df1[['d', 'date']], on='d')

# TASK 2 - Clustering

# TASK 3 - Sales Forecasting

# TASK 4 - Store Replenishment Use Case (MLOps)

# TASK FINAL - Evaluation design